File: mobile_new/lib/shared/utils/verrou_pari.dart
/// Jusqu'à quand propose-t-on d'enregistrer une mise ?
///
/// Miroir de `backend/src/services/verrou_pari.ts`. Le serveur refuse ; cet
/// écran ne doit pas proposer ce qui sera refusé — un bouton qui mène à une
/// erreur est pire qu'un bouton absent.
///
/// ── Ce que l'écran laissait faire ─────────────────────────────────────────
///
/// La carte d'accueil masquait le bouton « Miser » sur `status == 'finished'`
/// seulement. Pendant un match **en direct**, avec le score affiché juste à
/// côté, le bouton restait là. Le serveur, lui, ne vérifiait rien du tout : la
/// mise était enregistrée.
///
/// La page détail, elle, était déjà correcte — elle n'affiche le bouton que
/// sur `MatchStatus.upcoming`. C'est exactement le motif qui revient : une
/// règle appliquée à un endroit, oubliée à l'autre.
///
/// ── Pourquoi la date compte autant que le statut ──────────────────────────
///
/// `status` vient d'une synchronisation périodique côté serveur. Entre le coup
/// d'envoi et la synchronisation suivante, un match commencé porte encore
/// « à venir ». L'heure de coup d'envoi ne se périme pas.
///
/// Un banc compare cette règle à celle du backend sur les mêmes cas ; si l'une
/// des deux change seule, il tombe.
library;

/// Statuts qui signifient que le match n'est plus à venir.
const _plusAVenir = {'live', 'finished', 'suspended', 'postponed'};

/// Peut-on encore proposer d'enregistrer une mise sur ce pronostic ?
///
/// [resultat] est `null` tant que le pronostic n'est pas réglé.
/// [statutMatch] arrive tel que l'API l'écrit, dans n'importe quelle casse.
/// [dateMatch] est l'heure du coup d'envoi ; `null` si l'API ne l'a pas donnée.
bool pariEncoreOuvert({
  required String? resultat,
  required String? statutMatch,
  required DateTime? dateMatch,
  DateTime? maintenant,
}) {
  if (resultat != null && resultat.isNotEmpty) return false;

  if (_plusAVenir.contains((statutMatch ?? '').toLowerCase())) return false;

  if (dateMatch != null) {
    final t = maintenant ?? DateTime.now();
    if (!dateMatch.isAfter(t)) return false;
  }

  return true;
}

File: backend/src/services/verrou_pari.ts
/**
 * Jusqu'à quand un pari peut-il être enregistré dans la bankroll ?
 *
 * ── Ce qui n'était vérifié nulle part ──────────────────────────────────────
 *
 * `placeBet` contrôlait le solde, le montant et les doublons. Pas le match.
 * On pouvait donc enregistrer une mise sur un pronostic **déjà réglé**, dont
 * le résultat est public — et rien ne distinguait ensuite cette ligne d'un
 * pari posé avant le coup d'envoi.
 *
 * Ce n'est pas qu'une faille d'API. L'écran masquait le bouton « Miser » sur
 * `status == 'finished'` seulement : pendant un match **en direct**, avec un
 * score déjà affiché, le bouton restait là. Le parcours normal de
 * l'application permettait de miser en connaissant une partie du résultat.
 *
 * ── Pourquoi cela compte au-delà d'un compte ──────────────────────────────
 *
 * La bankroll n'est pas un pari réel chez un bookmaker : c'est le suivi que
 * PronoWin tient. Mais ce suivi alimente les statistiques de l'utilisateur, et
 * surtout le **classement public**. Un historique où l'on peut ajouter après
 * coup les paris gagnants rend le classement invérifiable — et le classement
 * est l'argument qui donne du crédit au reste.
 *
 * ── La date fait autorité, pas seulement le statut ────────────────────────
 *
 * `status` est tenu à jour par une synchronisation périodique (30 s en direct,
 * 2 min sinon). Entre le coup d'envoi et la synchronisation suivante, un match
 * commencé porte encore `SCHEDULED`. Se fier au seul statut laisserait donc
 * une fenêtre ouverte à chaque match. L'heure de coup d'envoi, elle, est celle
 * que l'application affiche partout : c'est elle qui ferme la saisie.
 */

/** Pourquoi un pari est refusé. Aucun refus n'est silencieux. */
export type RefusPari =
  | 'resultat_connu'
  | 'match_commence'
  | 'match_reporte';

/** Statuts qui signifient que le match n'est plus à venir. */
const PLUS_A_VENIR = new Set(['live', 'finished', 'suspended']);

/** Ce qu'on dit à l'utilisateur pour chaque refus. */
export const MESSAGE_REFUS: Record<RefusPari, string> = {
  resultat_connu:
    'Ce pronostic est déjà réglé : son résultat est connu.',
  match_commence:
    'Le match a commencé : les mises ne sont plus enregistrées.',
  match_reporte:
    'Ce match est reporté. Les mises rouvriront quand une nouvelle date sera connue.',
};

/**
 * Le pari peut-il encore être enregistré ? `null` si oui, le motif sinon.
 *
 * @param resultat    résultat du pronostic (`WIN` / `LOSS` / `PUSH`), `null` tant
 *                    qu'il n'est pas réglé
 * @param statutMatch statut du match, dans la casse où il arrive
 * @param dateMatch   heure du coup d'envoi
 * @param maintenant  injectable pour les bancs
 */
export function refusDePari(params: {
  resultat:     string | null | undefined;
  statutMatch:  string | null | undefined;
  dateMatch:    Date | null | undefined;
  maintenant?:  Date;
}): RefusPari | null {
  const { resultat, statutMatch, dateMatch } = params;
  const maintenant = params.maintenant ?? new Date();

  // Un pronostic réglé n'a plus rien d'incertain, quel que soit l'état du match.
  if (resultat != null && resultat !== '') return 'resultat_connu';

  const statut = (statutMatch ?? '').toLowerCase();
  if (statut === 'postponed') return 'match_reporte';
  if (PLUS_A_VENIR.has(statut)) return 'match_commence';

  // Le statut peut être en retard d'une synchronisation ; l'heure, non.
  if (dateMatch != null && dateMatch.getTime() <= maintenant.getTime()) {
    return 'match_commence';
  }

  return null;
}
